File: client/src/Services/user-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../Models/User';
import { Follower } from '../Models/Follower';

@Injectable({
  providedIn: 'root',
})
export class UserService {

  private apiUrl = 'http://localhost:5283/user';
  constructor (private http: HttpClient) {}

  searchUser(input: string): Observable<User[]> {
    return this.http.get<User[]>(
      `${this.apiUrl}/search/${input}`,
      { withCredentials: true }
    );
  }

  getUser(username: string): Observable<User> {
    return this.http.get<User>(
      `${this.apiUrl}/${username}`,
      { withCredentials: true }
    );
  }

  followUser(followeeId: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/follow/${followeeId}`,{}, { withCredentials: true });
  }  

  unfollowUser(followeeId: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/unfollow/${followeeId}`, { withCredentials: true });
  }  

  getFollowStatus(followeeId: string): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/follow-status/${followeeId}`, { withCredentials: true });
  }  

  getFollowing(userId: string): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/following/${userId}`, { withCredentials: true });
  }  

  getFollowers(userId: string): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/followers/${userId}`, { withCredentials: true });
  }  
  
}

File: server/Services/UserService.cs
using server.Models;
using server.Repositories;
using server.Dto;

namespace server.Services {
    public class UserService: IUserService {
        private readonly IUserRepository _repository;

        public UserService(IUserRepository repository){
            _repository = repository;
        }

        public async Task<IEnumerable<User>> SearchUser(string input){
            return await _repository.SearchUser(input);
        }

        public async Task<User> GetUser(string username){
            var result = await _repository.GetUser(username);
            return result;
        }

        public async Task FollowUser(string followerId, string followeeId){
            await _repository.FollowUser(followerId, followeeId);
        }

        public async Task UnfollowUser(string followerId, string followeeId){
            await _repository.UnfollowUser(followerId, followeeId);
        }

        public async Task<bool> GetFollowStatus(string followerId, string followeeId){
            var result = await _repository.GetFollowStatus(followerId, followeeId);

            if (result==null){
                return false;
            }

            return true;
        }

        public async Task<IEnumerable<UserDto?>> GetFollowing(string userId){
            return await _repository.GetFollowing(userId);
        }

         public async Task<IEnumerable<UserDto?>> GetFollowers(string userId){
            return await _repository.GetFollowers(userId);
        }
    }
}